google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkProgressListener.java

```diff
@@ -0,0 +1,187 @@
+import com.intellij.openapi.progress.util.AbstractProgressIndicatorExBase;
+import com.intellij.openapi.util.Disposer;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Progress listener for {@link ManagedCloudSdkService} installation and update processes. Uses
+ * {@link BackgroundableProcessIndicator} to show progress while running managed SDK jobs. UI
+ * provides cancel option.
+ *
+ * <p>This class is not thread-safe and must be called from single thread (managed SDK job thread).
+ */
+class ManagedCloudSdkProgressListener implements ProgressListener {
+
+  private final ManagedCloudSdkService managedCloudSdkService;
+
+  private BackgroundableProcessIndicator progressIndicator;
+  private Backgroundable task;
+
+  private double totalWork;
+  private double lastUpdatedGlobalDoneFraction;
+
+  ManagedCloudSdkProgressListener(ManagedCloudSdkService managedCloudSdkService) {
+    this.managedCloudSdkService = managedCloudSdkService;
+  }
+
+  @Override
+  public void start(String message, long totalWork) {
+    this.totalWork = totalWork;
+
+    // initialize progress indicator first in UI thread before using it for updates.
+    this.progressIndicator = initProgressIndicator();
+
+    setProgressText(message);
+
+    if (totalWork == UNKNOWN) {
+      progressIndicator.setIndeterminate(true);
+    }
+  }
+
+  @Override
+  public void update(long workDone) {
+    if (workDone == UNKNOWN) {
+      progressIndicator.setIndeterminate(true);
+    } else {
+      progressIndicator.setIndeterminate(false);
+      lastUpdatedGlobalDoneFraction += workDone / totalWork;
+      progressIndicator.setFraction(lastUpdatedGlobalDoneFraction);
+    }
+  }
+
+  @Override
+  public void update(String message) {
+    setProgressText(message);
+  }
+
+  @Override
+  public void done() {
+    if (!progressIndicator.isFinished(task)) {
+      progressIndicator.finish(task);
+      progressIndicator.dispose();
+    }
+  }
+
+  @Override
+  public ProgressListener newChild(long allocation) {
+    return new ChildProgressListener(allocation);
+  }
+
+  private void setProgressText(String message) {
+    progressIndicator.setText(GctBundle.message("managedsdk.progress.message", message));
+  }
+
+  @VisibleForTesting
+  BackgroundableProcessIndicator initProgressIndicator() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // task info for progress indicator, required to init/finish properly.
+              task =
+                  new Backgroundable(
+                      null /* not project specific task */,
+                      GctBundle.message("managedsdk.notifications.title"),
+                      true /* cancellable */,
+                      PerformInBackgroundOption.ALWAYS_BACKGROUND) {
+                    @Override
+                    public void run(@NotNull ProgressIndicator indicator) {
+                      // will not be called in manual progress mode.
```

Comment:
just for me can you explain what this means please?

Reply:
Sure. So general need was to use background progress indicator without using `ProgressManager` calls so we could still use own thread control (futures, single thread per job etc.). `BackgroundableProcessIndicator` works properly only when passed in a `Task` (usually constructed in `ProgressManager`, which at some point calls `run()`. Progress indicator itself doesn't do anything with `Task`, so we had to override this unused `run()` here. `Task` is also required to finish indicator. The only way to find out about cancel is `addStateDelegate()` callback. Twisted API and twisted code!

Reply:
oh man. ok, maybe add just a bit more text on that comment then for the next reader?

Reply:
Great idea, I'll expand on comments here.

Reply:
done.